effects/letter-clock/data/palette_extend.py

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
+
+# File format of an extended palette is as follows:
+# first line contains bitmasks of lenses that
+# will be applied to the columns below.
+# Then each line should contain:
+# unique base colour in the first column that coresponds
+# to a colour in the image file;
+# in the second column a colour by which the first one will be replaced
+# in the final palette (can be the same);
+# then each column should contain a colour which will be matched
+# left to right to the masks (lenses) supplied in the first line
+# in order to put the colour into a matching position.
+
+from PIL import Image
+import argparse
+
+
+def read_palette(im, max_colours):
+    if im.mode != "P":
+        raise SystemExit("Only 8-bit images with palette supported.")
+
+    pal = im.getpalette()
+    colours = im.getextrema()[1] + 1
+
+    if pal is None:
+        raise SystemExit("Image has no palette!")
+    if colours > max_colours:
+        raise SystemExit(
+            "Image has {} colours, expected at most {}!".format(
+                colours, max_colours
+            )
+        )
+
+    r_pal = [num >> 4 for num in pal]
+    r_cmap = [r_pal[i * 3: (i + 1) * 3] for i in range(colours)]
+
+    return r_cmap
+
+
+def extend_palette(orig_cmap, path, max_colours):
+    assert len(orig_cmap) == len(
+        set([tuple(colour) for colour in orig_cmap])
+    ), "elements in cmap must be unique"
+    # We fill any possible holes in a new cmap with the background colour.
+    # This way we don't need to enforce complete palettes in txt file.
+    new_cmap = orig_cmap + [orig_cmap[0]] * (max_colours - len(orig_cmap))
+    with open(path) as f:
+        lens_tab = [int(lens, base=2) for lens in f.readline().split()]
+        for line in f:
+            colour_line = [
+                [
```

Comment:
Great opportunity to write small function that performs the task of constructing RGB color from string.